lightning/src/ln/channel.rs

```diff
@@ -3996,45 +4145,162 @@ impl<SP: Deref> Channel<SP> where
 	/// blocked.
 	#[allow(unused)]
 	pub fn signer_maybe_unblocked<L: Deref>(&mut self, logger: &L) -> SignerResumeUpdates where L::Target: Logger {
-		let commitment_update = if self.context.signer_pending_commitment_update {
-			self.get_last_commitment_update_for_send(logger).ok()
-		} else { None };
+		log_trace!(logger, "Signing unblocked in channel {} at sequence {}",
+			&self.context.channel_id(), self.context.cur_holder_commitment_transaction_number);
+
+		if self.context.signer_pending_commitment_point {
+			log_trace!(logger, "Attempting to update holder per-commitment point...");
+			self.context.update_holder_per_commitment_point(logger);
+		}
+
+		if self.context.signer_pending_released_secret {
+			log_trace!(logger, "Attempting to update holder commitment secret...");
+			self.context.update_holder_commitment_secret(logger);
+		}
+
+		if self.context.channel_state & (ChannelState::PeerDisconnected as u32) != 0 {
+			log_trace!(logger, "Peer is disconnected; no unblocked messages to send.");
+			return SignerResumeUpdates::default()
+		}
+
+		// Make sure that we honor any ordering requirements between the commitment update and revoke-and-ack.
+		let (commitment_update, raa) = match &self.context.resend_order {
```

Comment:
So... _no_ we don't care about the order they're *called in* but _we must ensure that the following hold_:

- If order is `CommitmentFirst` and the `revoke_and_ack` is available but the `commitment_update` is not, we must not broadcast the `revoke_and_ack`.
- If order is `RevokeAndACKFirst` and the `commitment_update` is available but the `revoke_and_ack` is not, we must not broadcast the `commitment_update`.

If we collapse this particular fragment and instead generate the CU and RAA without regard for order, some of the async signer tests will fail; in particular, those that unblock the signer in a way that causes one of the above to be violated. For example, if the order is `CommitmentFirst`, we can cause the signer to reveal the RAA first without revealing the CU.

This, combined with the fact that we clear the `signer_pending_*` bit, means that it is difficult to handle this in the ChannelManager, which admittedly feels like it might be a more appropriate place to do so.

But maybe there's a cleaner way to do this?